pkg/providers/tinkerbell/validator.go

```diff
@@ -0,0 +1,149 @@
+		return errors.New("tinkerbellMachineConfig osFamily for control plane is not set or is empty")
+	}
+
+	if tinkerbellClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef == nil {
+		return errors.New("must specify machineGroupRef for worker nodes")
+	}
+
+	workerNodeGroupMachineConfig := tinkerbellClusterSpec.firstWorkerMachineConfig()
+	if workerNodeGroupMachineConfig == nil {
+		return fmt.Errorf("cannot find tinkerbellMachineConfig %v for worker nodes", tinkerbellClusterSpec.Cluster.Spec.WorkerNodeGroupConfigurations[0].MachineGroupRef.Name)
+	}
+	if len(workerNodeGroupMachineConfig.Spec.OSFamily) <= 0 {
+		return errors.New("tinkerbellMachineConfig osFamily for worker nodes is not set or is empty")
+	}
+
+	// TODO: move this to api Cluster validations
+	if err := v.validateControlPlaneIp(tinkerbellClusterSpec.Cluster.Spec.ControlPlaneConfiguration.Endpoint.Host); err != nil {
+		return err
+	}
+
+	if controlPlaneMachineConfig.Spec.OSFamily != anywherev1.Bottlerocket && controlPlaneMachineConfig.Spec.OSFamily != anywherev1.Ubuntu {
+		return fmt.Errorf("control plane osFamily: %s is not supported, please use one of the following: %s, %s", controlPlaneMachineConfig.Spec.OSFamily, anywherev1.Bottlerocket, anywherev1.Ubuntu)
+	}
+
+	if workerNodeGroupMachineConfig.Spec.OSFamily != anywherev1.Bottlerocket && workerNodeGroupMachineConfig.Spec.OSFamily != anywherev1.Ubuntu {
+		return fmt.Errorf("worker node osFamily: %s is not supported, please use one of the following: %s, %s", workerNodeGroupMachineConfig.Spec.OSFamily, anywherev1.Bottlerocket, anywherev1.Ubuntu)
+	}
+
+	if controlPlaneMachineConfig.Spec.OSFamily != workerNodeGroupMachineConfig.Spec.OSFamily {
+		return errors.New("control plane and worker nodes must have the same osFamily specified")
+	}
+
+	for _, machineConfig := range tinkerbellClusterSpec.machineConfigsLookup {
+		if machineConfig.Namespace != tinkerbellClusterSpec.Cluster.Namespace {
+			return errors.New("TinkerbellMachineConfig and Cluster objects must have the same namespace specified")
+		}
+	}
+
+	if tinkerbellClusterSpec.datacenterConfig.Namespace != tinkerbellClusterSpec.Cluster.Namespace {
+		return errors.New("TinkerbellDatacenterConfig and Cluster objects must have the same namespace specified")
+	}
+
+	return nil
+}
+
+func (v *Validator) validateControlPlaneIp(ip string) error {
+	// check if controlPlaneEndpointIp is valid
+	parsedIp := net.ParseIP(ip)
+	if parsedIp == nil {
+		return fmt.Errorf("cluster controlPlaneConfiguration.Endpoint.Host is invalid: %s", ip)
+	}
+	return nil
+}
+
+func (v *Validator) validateTinkerbellIP(ctx context.Context, TinkerbellIP string) error {
+	if TinkerbellIP == "" {
+		return fmt.Errorf("tinkerbellIP is required")
+	}
+
+	parsedIp := net.ParseIP(TinkerbellIP)
+	if parsedIp == nil {
+		return fmt.Errorf("cluster tinkerbellDatacenterConfig.tinkerbellIP is invalid: %s", TinkerbellIP)
+	}
+
+	return nil
+}
```

Comment:
should just have a validateIP function to avoid code repetition... we could possibly even think of putting that utility function in a more generic place that all providers could use

Reply:
I've created validateIP function which checks if IP is valid or not for both controlplaneHostIP and tinkerbellIP. We can move this function to a more specific place https://github.com/aws/eks-anywhere/blob/main/pkg/api/v1alpha1/cluster.go#L164. I'll create a separate PR for that as we will have to change it for other providers as well.